### ISSUE TITLE
重回帰するやつ

### ISSUE THREAD

**Issue by kuniotoku** [kuniotoku/chatGPT#5]
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression

import numpy as np

class RemoveNegativeValues(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        return self
    
    def transform(self, X):
        return X[(X >= 0).all(axis=1)]

class ReplaceOutliersWithMode(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        self.modes_ = np.nanmode(X, axis=0)
        return self
    
    def transform(self, X):
        X_copy = X.copy()
        for i in range(X.shape[1]):
            mask = np.abs(X[:, i] - self.modes_[i]) > 3 * np.nanstd(X[:, i])
            X_copy[mask, i] = self.modes_[i]
        return X_copy

# パイプラインを作成
pipeline = Pipeline([
    ('remove_negative', RemoveNegativeValues()),  # マイナスの変数を持つデータを除外
    ('outlier_replacement', ReplaceOutliersWithMode()),  # 外れ値を最頻値で補完
    ('regression', LinearRegression())  # 重回帰分析
])

# データを用意
X_train = ...  # 学習データの特徴量
y_train = ...  # 学習データの目標値

# パイプラインで重回帰分析を実行
pipeline.fit(X_train, y_train)

**Comment by kuniotoku** [comment]
# 重回帰分析の係数を取得
coefficients = pipeline.named_steps['regression'].coef_

# 重回帰分析の標準誤差を取得
std_errors = np.sqrt(np.diagonal(np.linalg.inv(np.dot(X_train.T, X_train)))) * np.sqrt(pipeline.named_steps['regression'].residues_ / (X_train.shape[0] - X_train.shape[1]))

# 係数のt値を計算
t_values = coefficients / std_errors
print("t値:", t_values)
